src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Symbolic/SymbolicRegexThresholds.cs

```diff
@@ -22,12 +22,18 @@ internal static class SymbolicRegexThresholds
         /// processing starts out in DFA mode, even if we've previously triggered NFA mode for the same regex.
         /// We switch over into NFA mode the first time a given traversal (match operation) results in us needing
         /// to create a new node and the graph is already or newly beyond this threshold.
+        /// TODO: summarize this
+        /// this should be a very last resort action, going from DFA mode to NFA mode turns 500MB/s to 5MB/s
+        /// with an entirely different search-time algorithmic complexity
+        /// 100_000 isn't a really a high memory cost either,
+        /// ideally NFA mode should never be used, 1_000_000 is ok as well but it depends how much memory the user has
         /// </remarks>
-        internal const int NfaThreshold = 10_000;
+        internal const int NfaThreshold = 100_000;
```

Comment:
What is the memory impact of raising this from 10K to 100K?

@veanes, what are your thoughts on these limits, given your knowledge of where this will be used?

Reply:
The creation of new DFA states not seen earlier could cause the derivatives to become **too large** (essentially as sets of NFA states) and then cause a lot of memory consumption (and also delay the timeout to be detected). When DFA switches to NFA mode, the main reason for doing this should be detecting such large derivatives (representing sets of NFA states) not really the number of states that in most typical cases could safely be 100k. Currently there is no mechanism for detecting such situations. In other words, I think that the NfaThreshold measure is really a somewhat rough estimate so in most typical cases 100k is fine. (However, the limit 100k does obviously increase the worst-case constant overhead in NFA mode.) Ideally there should be a limit for detecting too large sets of NFA states (as derivatives) in combination with a large number of states seen so far that should trigger NFA mode. But this is a difficult decision.

The NFA limit was in some sense also related to handling of timeout during DFA state creation. We worked on this with Ian earlier today and passed the timeout detection to the function that constructs the new transition and does the timeout check before a new derivative is constructed (every time as opposed to after certain intervals of the input) (This does not affect the hot-loop that just reads the dfa state array).